### ISSUE TITLE
Investigation : Push Constants

### ISSUE THREAD

**Issue by shaoboyan** [gpuweb/gpuweb#75]
## Intro

In many situations, like dynamic indexing, developers may pass a **small amount** of **read-only** information to pipeline and **update them frequently**.

Metal, D3D12 and Vulkan all provide optimizations for such kind of usages.

It is reasonable to bring these features to WebGPU.

### Metal

In Metal, [SetBytes](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes)(), [SetVertexBytes](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515846-setvertexbytes)() and [SetFragmentBytes](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1516192-setfragmentbytes?language=objc)() APIs are potential candidates to support this feature.

Here are the related descriptions in the documents of Metal:

 > Using these methods is equivalent to creating a new MTLBuffer object from the specified data and then binding it to the compute function with setBuffer() | setVertexBuffer() | setFragmentBuffer() method.

These methods **avoid the possible performance overhead of creating an intermediary MTLBuffer Object**.

To use these APIs, developers need to call `SetBytes()` for compute function, call `SetVertexBytes()` for vertex function and `SetFragmentBytes()` for fragment function to ensure the special data buffers are set to corresponding shader stages.

Metal requests the data buffers set by these APIs **less than 4KB** and must be **one-time-use data buffer in each render encoder**. Metal **doesn’t support updating a subset of the values** in these special data buffers.

### D3D12

In D3D12, **Root Constants** is potential candidate to support this feature.

According to [MSDN](https://docs.microsoft.com/en-us/windows/desktop/direct3d12/root-signatures-overview), Root Constants are 

> - **inline 32-bit values** 
> - Applications can define root constants in the root signature, each as a set of 32-bit values. They appear in High Level Shading Language (HLSL) as a constant buffer.

Actually, in Microsoft’s [DirectX-Graphics-Sample](https://github.com/Microsoft/DirectX-Graphics-Samples/tree/master/Samples/Desktop/D3D12DynamicIndexing), root constants are used in dynamic indexing.

To use root constants, developers need to register root constants into root signatures through `InitAsConstant()` and set the visibilities of the root constants for each shader stage. After that, it is linked to the command lists. Developers can update any subset of values in root constants through APIs like `setGraphicsRoot32BitConstants()` by providing the slot number, memory address of new data and range.

Here are the [limitations](https://docs.microsoft.com/en-us/windows/desktop/direct3d12/using-constants-directly-in-the-root-signature) of root constants :

-   Each set of user constants is treated as **a scalar array of 32-bit values**, dynamically indexable and read-only from the shader.  Note that constant buffers for historical reasons **are viewed as sets of 4x32-bit values**.
    
-   **Arrays are not permitted** in constant buffers that get mapped onto root constants since dynamic indexing in the root signature space is not supported.
    
-   **The maximum size of a root signature is 64 DWORDs. Root constants cost 1 DWORD each**, since they are 32-bit values.

### Vulkan

In Vulkan, **Push Constants** is ideal candidate to support this feature.

According to Vulkan [SPEC](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#descriptorsets-sets):

> Push constants represent **a high speed path** to modify constant data in pipelines that is expected to **outperform memory-backed resource updates**. 

So push constants in Vulkan is **a high speed way** to handle **small amount** of **read-only data** which needs to **updated frequently**.

But in Vulkan, as is different from D3D12 and Metal, **all shader stages access the same push constants block**. This means, even if values in push constants are updated with one shader stage flag, all of the shader stages in the pipeline will be affected.

To use push constants in Vulkan, developers need to fill the information in `vkPipelineLayout` to describe the ranges of the push constants that the pipeline wants to access and their the shader stage visibility. After that, these ranges are allocated. Developer can update any subset of values in ranges of push constants through API `vkCmdPushConstants()` with shader stage bits, memory address of new data, offset in bytes and length of updated range in bytes.

Here are the limitations of Vulkan push constants:

-   Push constants has length limitation.
    
-   Vulkan [use std430 layout](https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt#L216) in GLSL shader by default. Push constants block in shader may be padded.

-   Push constant variables with same offset cannot have different values in different shader stages.

## Proposal
There are common requests in Metal, D3D12 and Vulkan to support this feature. And we propose push constants in WebGPU should follow these common requests :

-   Before each render pass, **all push constants need to be set to zero**. So we don't have to validate it was set before a shader uses push constatns. And it aligns to Metal’s behavior that special data buffers become invalid after the encoder is submitted to the queue.
    
-   The offset of push constants blocks in shader must be a multiple of 4, as required by both Vulkan and D3D12.

    
-   In WebGPU, we have two candidate types of push constants :
    

	-  The first is **each shader stage has their own push constant ranges**, which means variables with same offset can have different value in different shader stages.
    
	-   The second is **a single push constant range that’s shared between all shader stages**, which means variables with same offset can only have 1 value in different shader stages. **Each subset of the push constant range can have its own shader visibilities**.

	The main difference of these two proposals is **whether the variable with same offset can have different value in different shader stages**.
	
-   We propose a structure for developers to describe push constants range and require sequences of push constants in WebGPUPipelineLayout :
	``` javascript    
	dictionary Pushconstants {
            WebGPUShaderStageBit stages;
            uint32_t offset;
            uint32_t length;
	};
	dictionary WebGPUPipelineLayoutDescriptor {
            sequence<WebGPUBindingGroupLayout> bindGroupLayouts;
            sequence<Pushconstants> pushConstants;
	};
	```
	And two pipeline layouts are **defined to be “compatible for push constants” if they were created with identical push constant ranges**. It means push constants values can share between pipeline layouts that are compatible for push constants.
	
-   The proposed API to **update push constants** is
	``` javascript
	partial interface WebGPURenderPass {
	    void setPushConstants(WebGPUShaderStageBit stages, uint32_t offset, uint32_t count, 
                              ArrayBuffer data);
	}
	```
	It works for both proposed push constant well. For _each shader stage has their own push constant ranges_, `offset` means offset for each shader stage push constant range. For _a single push constants range that's shared between shader stages_, `offset` means a global unique push constants range offset.

**Comment by msiglreith** [comment]
Thanks for the investigation!

A few points from my side:

> Each set of user constants is treated as a scalar array of 32-bit values ..

How is this different from Vulkan here?

> .. dynamically indexable and read-only from the shader. 

As mentioned in the next point, it's not possible to use arrays.

> The offset of push constants blocks in shader must be a multiple of 4, as required by both Vulkan and D3D12.

Depends a bit on the specific approach used. In case of Vulkan's push constants the root constants in dx12 required additional restrictions to build valid constant buffers with the correct offsets.

And some followup questions:

* What's the point of the first candidate? Shouldn't push constants be equal for all specified shader stages?

* How to deal with overlapping push constants?

Finally, for the number of push constants I would set it for all backends to 128 bytes. Supported by all backends and kind of upper limit on dx12 considering descriptors sets and dynamic descriptors requiring also space in the root signature. In that case implementation details (dx12's root signature layout) doesn't need to be exposed in the public API surface.

**Comment by shaoboyan** [comment]
@msiglreith thx for your feedback! Here're some comments:

> How is this different from Vulkan here?

For Vulkan, if high level shader language is GLSL, the push constants is default set to **std430 layout. This layout won't rounded three-component vector in array or structure up to four-component vector**. This is  different from dx12 in root constants.  But developer can also set push constants block in GLSL to **std140 layout, which will rounded all three-component vectors up to four-component vector**. In this situation, Vulkan is the same as dx12.

> What's the point of the first candidate? Shouldn't push constants be equal for all specified shader stages?

For Metal and D3D12, it is possible for developers to **have several push constants alike resources(e.g. root constants in dx12)**. These resources can have different shader visibilities and contain different values. 
The first candidate is based on this.
However, I perfer second propose myself. For it is enough for scenario like dynamic indexing and quite easy to do bound checking and deal with overlapping.

> How to deal with overlapping push constants ? 

Good question. 
I want to deal with overlapping based on these rules : 

- First, **all of the shader stages that can read from the same push constant range must aware value changes in this push constant range**. This means, if value in the push constant range updated by any shader stage, all of the remain stages which can read from this range must see the new value.

- Second, I plan to introduce a [Vulkan alike](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#descriptorsets-sets) limitation, which is **any two elements in `sequence<Pushconstants>` cannot have the same WebGPUShaderStageBit value**. This will reduce fragmentation of push constant range and prevent situations that several push constant ranges ,which have same shader stage visibilities, overlap each other.

> Finally, for the number of push constants I would set it for all backends to 128 bytes.

Do you think add a API for developer to query the maximum push constant range is a good idea ?

>  ... kind of upper limit on dx12 considering descriptors sets and dynamic descriptors requiring also space in the root signature. In that case implementation details (dx12's root signature layout) doesn't need to be exposed in the public API surface.

I'm not quite understand this. Do you mean the max size of root constants is 64DWORD (256 bytes)? 

I've wrote a simple sample that creating a CRV in HLSL with 68 uint32 values which the back resource of them are root constants. This sample works fine, so I don't think 64 DWORD is the max size of root constants.

**Comment by kvark** [comment]
Thank you for the investigation!

Push constants look like an easy feature to have for us, since all the backends can efficiently support it. I don't see much controversy of exposing them (variant 2, where there is a single push constant array visible to all stages, is a natural choice). I mean, there is stuff to discuss wrt shader visibility flags, but we have to narrow this down anyway for resource bind groups, so push constants aren't making our work more difficult.

As for the maximum limit, we should just settle on a reasonable minimum, and then think about exposing higher limits after MVP.

**Comment by msiglreith** [comment]
Thanks for clarification.

> Do you think add a API for developer to query the maximum push constant range is a good idea ?

IMO hardcoding it to 128bytes would be fine.

> I'm not quite understand this. Do you mean the max size of root constants is 64DWORD (256 bytes)?

To transparently expose descriptor sets and push constants I'm suggesting to split the root signature statically:
* 32 DWORDs for root constants
* 4 descriptor tables (CBV_SRV_UAV) (8 DWORDs)
* 4 descriptor tables (Sampler) (8 DWORDs)

leaving 16 DWORDS open.
Depending on the feature set of gpuweb:
* 8 DWORDs for dynamic uniform buffers (but ugly to implement..)
* 4 DWORDs for dynamic storage buffers
* <4 DWORDs for internal usage (e.g shader builtins)>

else increase the descriptor set limit.
This fixed splitting makes it easier for users to manage their resources and it's supported by all backends.

**Comment by shaoboyan** [comment]
@msiglreith Thx for explaination.

**Comment by Kangz** [comment]
+1 for starting with 128bits I was going to suggest 64bits but that's a bit extreme.

@msiglreith's root signature split looks good, but I think in many cases they could be packed more because we know all the required information at WebGPUPipelineLayout creation time.

**Comment by jdashg** [comment]
@Kangz You mean 128bytes, yes?

**Comment by jdashg** [comment]
(Oops, misclicked!)

**Comment by Kangz** [comment]
I meant 128bits, ```4 x uint32_t``` just to be extra opinionated and leave space for implementation push constants.

**Comment by Kangz** [comment]
@kvark to be opinionated that it should be used for indices into arrays in uniform buffers and the like and not for whole matrices. Also leaves space for the WebGPU implementation to use push constants for some of its own stuff.

**Comment by kvark** [comment]
What's conceptually a problem with the user using push constants for matrices if they need to be updated every draw call and they fit within limits?

> Also leaves space for the WebGPU implementation to use push constants for some of its own stuff.

That's a tough one, and it's a good point. In order to put any sort of a limit here we need an upper bound on how many push constants are needed internally. At the same time, if we don't introduce the limit, then we'll have more work for MVP in terms of implementation, since we'd need a fallback path on all backends.

**Comment by kainino0x** [comment]
I'm in favor of making the push constant space very small in MVP, so we can use the rest of the space for implementation. (128 bits = 16 bytes = 8 u16s = 4 u32s seems ok, maybe we'll want to expand that a bit later).

**Comment by kvark** [comment]
@kainino0x that position is understandable. But how do you come up with a "very small" metric that is safe w.r.t. our future requirements?

**Comment by kainino0x** [comment]
I'm not sure. It just gives us more freedom if we start small now and increase later when we have a better idea of how much we might be using in implementations.

**Comment by litherum** [comment]
D3D12 shares the space used for root constants with everything else in the root signature. This means that if the Web programmer uses many descriptor tables (descriptor sets), she could run out of space for root constants. If we pick a fixed number of push constants to allow, the only way this could work is if we impose similar limits on the number of descriptor sets, or we shuffle around all the descriptors at pipeline creation time to coalesce them into a fewer number of sets.

Vulkan [requires](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#features-limits-required) that every implementation must have at least 128 bytes of push constants. It seems anything larger than that would mean we would lose compatibility with devices. I wonder if there are any Vulkan implementations that don't support push constants in hardware, but emulate them with a ring-buffer of buffers in order to claim conformance with this requirement.

128 bytes is small enough that it could be difficult to use, especially when there are other methods to upload a somewhat-unbounded amount of data to the GPU (using the ring-buffer method described above). Automatically promoting data to push constants is likely impossible, but it could be done with a hint (like the one proposed in this thread!) from the author, and could gracefully degrade to use buffers transparently. This has the benefit that the implementation can use as many push constants as the device supports, and not be limited by the minimum spec requirement of Vulkan. Programmers shouldn't have to opt-in to good performance. As described above, the implementation will likely already be shuffling descriptors around.

Therefore, I'm proposing accepting this API (provided all the other unrelated feedback is addressed), but not limiting the number of push constants in the pipeline layout beyond the length limits of regular buffers. We can add a normative requirement in the spec that if the number of push constants is lower than the threshold, the implementation must use push constants / root constants (but not providing any guarantees on the performance impact this would have).

Some other feedback:
- It is unlikely that the same set of constant data will be valuable in both the vertex shader and the fragment shader of the same draw call. Therefore, push constants in one stage should not be visible in the other stage, since making them visible in both stages would likely be a waste of space. If a user wants them available in both stages, the user can set them twice.
- Just like Direct3D, we should model this as a sequence of uint32s. We should address them by index, not by byte-offset. This removes the alignment requirements.
- The proposal mentions "compatible for push constants" but I don't understand what this term is used for. If a WebGPU programmer knows that two pipeline layouts are compatible, what does that get them in WebGPU? (I know what it means in Vulkan, but I don't know what it means in WebGPU.)
- Metal's [setBytes()](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes) does not allow updating a portion of a set of push constants. We can accommodate this in one of two ways:
  - Give each shader stage a single array of constants, and only allow updating the entire array
  - Give each shader stage a collection of arrays of constants, and each item in the collection is identified uniquely. The API allows updating entire items in the collection.
- This provides a natural solution to the "overlapping ranges" problem.
- One thing that's missing from this investigation is performance findings. The purpose of root constants are to improve performance; how much faster are they? We shouldn't add features simply because other APIs have them; we should add features because they have important use cases. I'm happy to gather this data (unless someone else wants to).
- Nit: some of these names are bad. The capitalization is bad in `dictionary Pushconstants`. I don't know what these things have to do with "pushing;" I propose just calling them "constants."

**Comment by kainino0x** [comment]
I don't think "constants" is descriptive enough. Specifically, they are constants accessible from shader programs (programmable stage invocations). I've sort of assumed they're like "pushing" onto the stack before the main function invocation (like args to main).

I have more comments but not time to write them now.

**Comment by kvark** [comment]
@m

**Comment by Kangz** [comment]
Making push constants unbounded with the assumption that they can fallback to a UBO is very worrisome to me. The problem comes from the fact that Vulkan has a minimum "maxBoundDescriptorSets" of 4 and requires that the descriptor set layout used by the pipeline (through the pipeline layout) is equal by value to the layout used to create the descriptor set.

More precisely, if we had a pipeline that has too many push constants and needs to fallback to a uniform buffer, where do we put the binding for the UBO? If the pipeline uses less than 4 bindgroups, then we can use one to store just the push constant UBO (PCUBO). If the 4 bindgroups are used, then it becomes hell: we need to modify the layout of one of them to accomodate the PCUBO. This means that for each bindgroup the application creates, they may or may not need an extra UBO that's not known in advance. This means we have to create descriptor sets on the fly as we are generating Vulkan commands and can't really re-use them across frames.

**Comment by zeux** [comment]
Something that came up in this topic in two different contexts is restrictions on the size of RS in D3D12 (64 DWORDs) and push constants in Vulkan (128 bytes). I'd like to caution against assuming that because the API has a certain limit, going up to this limit is free/cheap.

In particular, on one architecture that we have public information about (AMD), the "real" hardware has a moral equivalent of a root signature that can contain constants and some descriptors. The root signature is specified per shader stage when scheduling shader workloads, and the size that can be efficiently communicated to each stage is 16 DWORDs (64 bytes). In practice, a large portion of this will be occupied by constant buffer descriptors or pointers to descriptor tables. If you use a larger root signature, or large sets of push constants, you will force the driver to spill the descriptor or constant data to an internally managed ring buffer. This isn't a very fast path - while in theory the driver should be able to efficiently perform the ring buffer updates, in practice this code is not always optimal.

So on AMD if your goal is ultimate performance, the use cases for push constants are limited to a few integers per stage that specify offsets into other descriptor tables. For example, maybe you have a giant storage buffer that specifies material parameters - colors, texture indices, etc. - and you use an integer to specify an offset into this array (material id). That is most efficiently communicated via a push constant since it eliminates *all* indirections on the GPU side associated with fetching this data (it's faster than a dynamic uniform buffer which requires 1 shader indirection, and faster still than a non-dynamic uniform buffer which requires 2).

In some cases you need to specify several integers, and some of them might be required by just one stage, and some - by several. For example, you'd pass the vertex data offset to VS, and the material index to both VS & FS. I'm not sure how important it is to split push constants between stages since ultimately you can afford very few before you spill anyhow.

I can't comment on non-AMD architectures but my understanding has been that they either just don't support push constants in hardware (= require driver-managed ring buffer emulation in all cases), or do have special constant memory that is big enough to fit 128 bytes worth of data.

**Comment by zeux** [comment]
Oh - also - Metal setBytes is a pretty different beast. It's implementation is up to the driver of course, but I believe it's somewhat more akin to vkCmdUpdateBuffer in the sense that the constant data can be embedded into the command buffer and read by the GPU (vkCmdUpdateBuffer implies a copy to a buffer with potentially unbounded lifetime whereas setBytes implies that once GPU is done processing the command buffer, the memory could be reclaimed - a typical implementation would involve setting up GPU uniform buffer registers to point directly to the chunk of constant data that's allocated with a similar policy to command buffer (same lifetime, same memory management, limited maximum size of a page - hence the limit on size specified in setBytes). So it's useful to think of setBytes as a simple way to specify a relatively large amount of input data for a shader without having to manage buffer lifetime yourself as opposed to exercising a dedicated hardware resource built for small amounts of constant data.

**Comment by litherum** [comment]
> embedded into the command buffer

nope, it's a ring buffer.

**Comment by zeux** [comment]
@litherum Sure - details. The ring buffer fundamentally has to be synchronized with the command buffer wrt lifetime so you can think of it as being separate or reusing the same pages, it doesn't matter that much in the end.

**Comment by litherum** [comment]
I did a little bit of performance gathering on D3D12 on a 3-year Nvidia discrete GPU

![screen shot 2018-10-29 at 3 16 13 pm](https://user-images.githubusercontent.com/918903/47683725-b4809f00-db8d-11e8-942b-3bf092a827b6.png)

The shader:
```hlsl
cbuffer Constants : register(b0, space0)
{
	float i0;
	float i1;
	float i2;
	float i3;
	float i4;
	float i5;
	float i6;
	float i7;
};

float4 main(PixelShaderInput input, uint primitiveID : SV_PrimitiveID) : SV_TARGET
{
	float result = 0;
	for (int i = 0; i < 100000; ++i) {
		result += i0;
		result += i1;
		result += i2;
		result += i3;
		result += i4;
		result += i5;
		result += i6;
		result += i7;
	}
	return float4(result, result, result, 1.0f);
}
```

**Comment by litherum** [comment]
Here are the results from my Surface Book:

Intel GPU:
<img width="815" alt="screen shot 2018-10-29 at 10 40 47 pm" src="https://user-images.githubusercontent.com/918903/47698264-e7498800-dbcb-11e8-83dc-c6e0bdd3de9f.png">

Nvidia GPU:
<img width="880" alt="screen shot 2018-10-29 at 10 40 34 pm" src="https://user-images.githubusercontent.com/918903/47698284-f4ff0d80-dbcb-11e8-9e73-4768d5f05b47.png">

**Comment by kvark** [comment]
@litherum thank you for performing the benchmarks!

Can you make sense of the data? It's hard to see the difference from the graphs, although there does appear to be one.

I'm also thinking that your test is benchmarking a different aspect from what I had in mind. Going 100000 iterations in the shader means you are trying to see how performant is reading the constants, is that right? I think it would be more important to benchmark how performant is *passing* the constants, i.e. having a 100000 draw calls where the constants are changed for each.

**Comment by litherum** [comment]
> @litherum thank you for performing the benchmarks!
> 
> Can you make sense of the data? It's hard to see the difference from the graphs, although there does appear to be one.

The data looks identical to me. I’ll run a T-test to see what it says.

> 
> I'm also thinking that your test is benchmarking a different aspect from what I had in mind. Going 100000 iterations in the shader means you are trying to see how performant is reading the constants, is that right?

That’s right. I was under the impression that the whole point of root constants was that they were faster to read inside a shader.

> I think it would be more important to benchmark how performant is _passing_ the constants, i.e. having a 100000 draw calls where the constants are changed for each.

Sure, should be a simple modification.

**Comment by kvark** [comment]
> Welch's T-Test says

Ok, thanks! That sounds like an advantage not worthy pursuing.

> Sure. The benchmark now measures GPU time, but if I make this modification, I should be measuring CPU time, right?

It would measure CPU + transfer time.

Sounds like we understood the core promise of push constants very differently :)

**Comment by zeux** [comment]
Here's what my expectations for push constants would be:

- For small amounts of data, it can be faster than the next fastest alternative wrt CPU time - which in Vulkan would be rebinding the descriptor set with different dynamic offsets - because of a more direct driver update path
- For large amounts of small drawcalls, it can be faster than the next fastest alternative wrt GPU time because push constants can remove an extra indirection on GPU

I would expect that if you have small amount of draw calls with a very expensive shader, then the indirection won't matter because you'd be using the GPU constant cache in either case, and the CPU time obviously won't matter since you have only a few draw calls. Unfortunately I don't have readily available numbers to share but I would suggest measuring two workloads:

1. A lot of draw calls, each drawcall draws one point, reads point color from push constants. Measure CPU & GPU time.
2. A few draw calls, each draw call draws a lot of points, reads point color (shared between all points) from push constants. Measure GPU time.

Ideally the measurements would be done in Vulkan on AMD & NVidia GPUs - they have substantially different driver and GPU architectures.

**Comment by bbernhar** [comment]
On Intel GPUs, the API-defined max push constant size tends to be much smaller than what hardware allows before spilling into a buffer. The limit is near a couple KB (for all shader stages) and that'll likely get larger.

**Comment by devshgraphicsprogramming** [comment]
**You existing benchmark**

> For large amounts of light warps/wavefronts, it can be faster than the next fastest alternative wrt GPU time because push constants can remove an extra indirection on GPU

You are actually not supposed to notice any difference in shader speed between using Constant (Uniform) Buffers or Root (Push) Constants.

The difference is all in the CPU+driver cost of both updating and switching between Buffer binding offsets (even if you keep buffer object the same) vs feeding push_constants.

>I wonder if there are any Vulkan implementations that don't support push constants in hardware, but emulate them with a ring-buffer of buffers in order to claim conformance with this requirement.

Depends what you mean by "in hardware" because device local constant buffers without dynamic offsets sound pretty hardware to me (after all one cannot dynamically index registers).

In many architectures (possibly NVidia, but they give no details) your Constant Buffer values and Root Constants will be fetched in exact same way for the shader on the GPU side.
Again AMD is special where it manages to abuse a few of its registers for push constants that were originally meant for storing descriptor sets and bound buffer pointers , but also they have that funny case where UBO and SSBO fetches go almost through the same path.

> dynamically indexable
I think as soon as AMD driver detects D3D12 shaders that dynamically index Root Constants, it will stop doing that register optimization.

Furthermore, even on architectures where that is not the case and push constants already sit in special registers when your invocations are launched, bare in mind that the very first invocations and their data loads from the Constant Buffers will immediately put the Constant Buffer data in GPU core's L1 cache, where it will stay until eviction.
Good shader compilers (NVidia) will even emit wave-intrinsics to only have one invocation issue the read and then broadcast it to other invocations without even involving the L1 or shared memory.
I don't know whether GPUs with ARB_ballot but without wave-intrinsics use shared memory (which is always there in hardware, even in fragment shaders) to do a similar broadcast.

So in the above light, the following benchmark would only measure on-chip-shared/L1 vs. register read speed.
> A few draw calls, each draw call draws a lot of points, reads point color (shared between all points) from push constants. Measure GPU time.
Any difference in performance will get dwarfed by relative rasterization and vertex processing time.

This is probably why you are seeing identical timings.

So your benchmark should only measure CPU timings  not GPU timings.

We actually did a benchmark (albeit in OpenGL) for which buffer memory type is best for uniform buffers, which included both CPU and GPU timings:
https://github.com/buildaworldnet/IrrlichtBAW/wiki/Best-IGPUBuffer-Types-for-UBO-data-sourcing

Not entirely related, I know but its a pretty useful bit of information in case one would want to emulate push_constants.

Naturally Device local + Immutable (API un-updatable) + Unmappable is best.
(Actually updated via staging buffer and API buffer-2-buffer copies).

Combine this with other benchmarks available on the internet about Push_Constant vs. UBO and you will get the true cost of UBOs.


**Onto the 128-byte limit**

This is a very good limit, and there is a reason why Intel sets it so high, which is it doesn't have to copy the data you provide 2 times over (argument to command buffer, command buffer to device). No aforementioned ring buffers in the background.

As mentioned before with the ring buffers.
Other vendors the driver will have to manage the staging of these updates by itself, doing a copy to mapped memory behind the scenes as well as a DMA transfer to device (only AMD has a device local, yet mappable heap).

>What's conceptually a problem with the user using push constants for matrices if they need to be updated every draw call and they fit within limits?

>128 bytes is small enough that it could be difficult to use

The limit is not bad because Root (Push) Constants are intended to feed very very fast changing (per-draw) variables like:
- BaseInstanceID
- ObjectID
- MaterialID

And not your actual transform matrices and material parameters which should be in UBOs that are updated once per-frame/projection (and hopefully bound once too).

There per-draw IDs are essentially the only way to get "selector indices" that change between draw-calls, without which you'd have to bind UBOs at different offsets to get some variable that changes between draw-calls.

> Therefore, I'm proposing accepting this API (provided all the other unrelated feedback is addressed), but not limiting the number of push constants in the pipeline layout beyond the length limits of regular buffers. We can add a normative requirement in the spec that if the number of push constants is lower than the threshold, the implementation must use push constants / root constants (but not providing any guarantees on the performance impact this would have).

>> Making push constants unbounded with the assumption that they can fallback to a UBO is very worrisome to me. 

That will be royal mess full of bugs in browser implementations, also would give very non-portable performance (update buffers + transient constant buffers overflowing because user thought push_constants are convenient and used them everywhere).

> Finally, for the number of push constants I would set it for all backends to 128 bytes.

This is a very good idea, because push_constants were designed to have a small maximum size so the API would not have to do a lot of extra memory allocation for them (you can always afford to waste 128 bytes per-draw-call per-frame).

** Using Different Push Constant data for Different Shared Stages **

>In some cases you need to specify several integers, and some of them might be required by just one stage, and some - by several. For example, you'd pass the vertex data offset to VS, and the material index to both VS & FS. I'm not sure how important it is to split push constants between stages since ultimately you can afford very few before you spill anyhow.

I think if one API does not support different push (root) constants per-shader-stage then webGPU shouldn't either.

Also 128bytes is so little that you may as well not bother with updating sub-ranges, just take a `length` parameter that is a multiple of 4 so you don't always have to update them when `length==0`.

** Other Things **

> Automatically promoting data to push constants is likely impossible, but it could be done with a hint (like the one proposed in this thread!) from the author, and could gracefully degrade to use buffers transparently. 

Hints and implicit behaviour are one of the reasons why programmers hate OpenGL its derivatives (WebGL included).

> This layout won't rounded three-component vector in array or structure up to four-component vector. 

Uniform block still has to follow alignment rules for all other types, so if you place a `vec3` between some `vec4`s then it will round up.
It will also round up arrays of `vec3`.
Generally speaking vec3 is a mess to use, you should always place a float or an (u)int after it, even as padding, to sleep well at night.


TL,DR; If Vulkan gives same push_constants to all shaders then make webGPU have same constants for all shaders, make the push_constant size a hard API limit of 128 bytes, do not play around with updating push_constant ranges and only support one range that starts at 0 and goes up to `length` (mult of 4) bytes

**Comment by devshgraphicsprogramming** [comment]
I have examined the Dec 3 meeting minutes, you seem to think that you're faced with a choice between dynamic buffer offsets and root constants.

This is completely not the case, you NEED both, they are complementary.

**Comment by kainino0x** [comment]
Things are still fuzzy but I think the current thoughts are between:
- Dynamic buffer offsets
- Dynamic buffer offsets + "large push constants" (may not fit in root constants; may overflow and become a uniform buffer)
- "Small push constants" (root constants)

**Comment by devshgraphicsprogramming** [comment]
Well in case I wasn't clear I was advocating for "Dynamic buffer offsets + small push constants" which make most sense since these are the explicit mechanisms exposed by all 3 APIs.

Small push constants will not do the job of dynamic buffer offsets, you can think of them as indices into an array while dynamic offsets are more like fast-pointers.

**Comment by litherum** [comment]
I updated the [benchmark](https://github.com/litherum/PushConstantsBenchmark) to measure CPU performance, as described here. Here is the pseudocode for the [render function](https://github.com/litherum/PushConstantsBenchmark/blob/master/RootConstantPerformance/Content/Sample3DSceneRenderer.cpp#L542):

```C++
m_commandList->SetPipelineState(m_pipelineStateRootConstants.Get());
...
for (UINT i = 0; i < 1000; ++i) {
    float data[8] = { ... };
    m_commandList->SetGraphicsRoot32BitConstants(1, 8, data, 0);
    m_commandList->DrawIndexedInstanced(36, 1, 0, 0, 0);
}

m_commandList->SetPipelineState(m_pipelineStateBufferConstants.Get());
...
for (UINT i = 0; i < 1000; ++i) {
    UINT8* destination = ...;
    float data[8] = { ... };
    memcpy(destination, &data, sizeof(data));
    m_commandList->SetGraphicsRootConstantBufferView(1, ...);
    m_commandList->DrawIndexedInstanced(36, 1, 0, 0, 0);
}
```



I ran it on the same 3 configurations as above (because those are the ones I have access to right now). The results are timing how long CPU-side functions took, using `QueryPerformanceCounter()`; lower is better. The units are microseconds. "Upload" is the time the `memcpy()` took, note how it is always 0 for root constants. "Set" is `SetGraphicsRoot32BitConstants()` or `SetGraphicsRootConstantBufferView()`. "Draw" is `DrawIndexedInstanced()`. Here are the raw results: [Root Constants and Buffer Constants.zip](https://github.com/gpuweb/gpuweb/files/2660102/Root.Constants.and.Buffer.Constants.zip)

Desktop Nvidia GPU shows root constants are a 7.5% progression over buffer constants:
![screen shot 2018-12-08 at 3 05 29 pm](https://user-images.githubusercontent.com/918903/49691578-c47b8f00-fafa-11e8-9f6a-8547f95f5deb.png)

Laptop Intel GPU shows root constants are a 1.9% progression over buffer constants:
![screen shot 2018-12-08 at 3 06 31 pm](https://user-images.githubusercontent.com/918903/49691584-dceba980-fafa-11e8-8020-ab7895574d0b.png)

Laptop Nvidia GPU shows root constants are a 18.0% regression from buffer constants:
![screen shot 2018-12-08 at 3 07 07 pm](https://user-images.githubusercontent.com/918903/49691590-f260d380-fafa-11e8-8cb2-a3e7cad1df83.png)

**Comment by litherum** [comment]
> > @litherum thank you for performing the benchmarks!
> > Can you make sense of the data? It's hard to see the difference from the graphs, although there does appear to be one.
> 
> Welch's T-Test says: The 3 year old Nvidia GPU shows root constants are a statistically-significant progression of 1.3%, and the Surface Book's Nvidia GPU shows root constants a statistically-significant progression of 0.52%. The Intel GPU's results are not statistically-significant.
> 
> (The P-Values are 1.0350673044932595e-18 for the 3 year old Nvidia GPU, 0.6784765015669515 for the Intel GPU, and 6.16656053788606e-05 for the Surface Book's Nvidia GPU.)
> 

Oh, it seems like I never actually commented here about my previous GPU timings; I only mentioned it verbally in the 2018-11-12 call. Turns out the findings in 
https://github.com/gpuweb/gpuweb/issues/75#issuecomment-434101800 and 
https://github.com/gpuweb/gpuweb/issues/75#issuecomment-434179006 are bogus. The benchmark had two modes: one to test root constants, and one to test buffer constants, and so I would run the benchmark twice to gather the data, once in each mode. The problem was that by the time the first run finished, the device would be hotter, and would throttle, making the second run artificially slower. If I switched the order of the runs, I would get opposite results.

I updated that GPU-timings benchmark to interleave the two tests, and gathered new results: the NVIDIA GPUs had no statistically significant difference between root constants and buffers, and the Intel GPU had a statistically significant regression to use root constants.

This latest benchmark I just posted also has the interleaved behavior, of course.

**Comment by grorg** [comment]
This was discussed in the [10 Dec 2018 meeting](https://docs.google.com/document/d/1rGGgAwGuzWJ4i3fmgOHtaU1xkAQEiM7DZGE5V798Ga4)

**Comment by devshgraphicsprogramming** [comment]
```
m_commandList->SetPipelineState(m_pipelineStateBufferConstants.Get());
...
for (UINT i = 0; i < 1000; ++i) {
    UINT8* destination = ...;
    float data[8] = { ... };
    memcpy(destination, &data, sizeof(data));
    m_commandList->SetGraphicsRootConstantBufferView(1, ...);
    m_commandList->DrawIndexedInstanced(36, 1, 0, 0, 0);
}
```

Why are you appearing to  step on your own constant (uniform) buffer that's in flight?

Also you are rebuilding a command list (buffer) in its entirety each frame, thats not what the API was designed for, nor root constants.
https://github.com/litherum/PushConstantsBenchmark/blob/master/RootConstantPerformance/Content/Sample3DSceneRenderer.cpp#L586

Your benchmark might actually spend more time trying to get GPU timestamps than actually drawing... the only way to benchmark this is to measure a time difference (on CPU and GPU) between 100s of thousands of uninterrupted drawcalls without extra PIX debug markers, queries etc.
(Queries actually come from pools and are expensive)

**Comment by kvark** [comment]
> Why are you appearing to step on your own constant (uniform) buffer that's in flight? Where are the fences/events?

The offsets are computed based on the frame id. As far as I see, there is no stomping on the existing data.

> Also you are rebuilding a command list (buffer) in its entirety each frame, thats not what the API was designed for, nor root constants.

Currently, WebGPU doesn't have re-usable command buffers, so the test is for primary (one-shot) use case.

> Your benchmark might actually spend more time trying to get GPU timestamps than actually drawing

I'll try to test if this is the case.

**Comment by devshgraphicsprogramming** [comment]
I speculate that the main bottleneck to push_constant CPU speed in this benchmark is the command-buffer memory allocation and growth, with the UBO buffer you've already allocated all the memory for the small per-object data that you need. With push_constant/RootConstant you have a problem of a std::vector<>-like object inside a command buffer that has to grow each time you push some more RootConstants.


Also
> Currently, WebGPU doesn't have re-usable command buffers, so the test is for primary (one-shot) use case.

The whole point of recording things into a command buffer such as small buffer updates and root constants is so that this constant cost is amortized across runs.

If you re-use command buffers (usually secondary command-buffers) then the whole cost of allocating the space for the pushing of RootConstants is a one-time payment.
Vulkan also has an option to "recycle" command buffers via a `reset`.


The way it would usually work is that you record a secondary command buffer (or cmd list) that has all of your object draw calls for within a renderpass, and then re-use that static command buffer within a primary cmd buffer that gets generated on the fly.
So that way a secondary command buffer holds all the push-constants that can specify things such as object IDs, static world transforms, material parameters, etc.

Any culling is then done via DrawIndirect or by having multiple secondary command buffers corresponding to clusters of objects in a culling hierarchy.

**Comment by litherum** [comment]
> then re-use that static command buffer

Metal doesn’t have any facilities for re-usable command buffers, and we don’t want to require implementations to maintain their own copy of command buffer contents. This is why WebGPU, and therefore this benchmark, doesn’t have re-usable command buffers.

**Comment by devshgraphicsprogramming** [comment]
Well in that case I'm surprised that Root Constants are still slower despite you calling a reset.

**Comment by litherum** [comment]
I'm not sure that the only reaction to this data is "abandon push constants." The evidence also seems to support the "support soft root constants that gracefully degrade into a ring buffer" alternative.

**Comment by kvark** [comment]
@litherum 
> support soft root constants that gracefully degrade into a ring buffer

I'd like to see this happen. It'd be good to have an option to switch to the real push constants on some platforms at some time later down the road.

**Comment by devshgraphicsprogramming** [comment]
> The evidence also seems to support the "support soft root constants that gracefully degrade into a ring buffer" alternative.

That's something only a graphics middleware should do, I would rather have real push constants as an option or no push constants at all.

But obviously dynamic buffer offsets are mandatory to be supported.

**Comment by Kangz** [comment]
> support soft root constants that gracefully degrade into a ring buffer

See my worry about this in https://github.com/gpuweb/gpuweb/issues/75#issuecomment-430643222

**Comment by shaoboyan** [comment]
Notes: This won't include in WebGPU yet.

**Comment by aloucks** [comment]
Push constants would still be useful even with a limit of 128 bytes. UBOs with dynamic offsets can be wasteful due to the alignment restrictions when you just need to update a handful of indices or other small amounts of data.

**Comment by sseanwang** [comment]
Any update on push constants? It is quite useful in a large scene which contains millions of objects. Dynamic uniform just kind of wasteful and use too much memory due to the alignment restrictions.

**Comment by Kangz** [comment]
Push constants are not in scope for v1 for sure. They might be something that's added to WebGPU after v1 based on feedback from folks. (personally I think it is likely, but that the push constant space will be **very** limited, for example 16bytes, so you can pass indices to stuff but not whole matrices)

**Comment by sseanwang** [comment]
I think 128 byte which is the minimal support size in vulkan is a perfect size, and it will make things more flexiable and open more posibilities in rendering stratagy.

**Comment by Kangz** [comment]
I don't think WebGPU will be able to do 128 bytes because implementations will want to reserve push constants for their own workarounds injected in shaders. But also root signature space in D3D12 is very limited and the maximum amount of bindings is already consuming a lot of it.

**Comment by devshgraphicsprogramming** [comment]
> Push constants would still be useful even with a limit of 128 bytes. UBOs with dynamic offsets can be wasteful due to the alignment restrictions when you just need to update a handful of indices or other small amounts of data.

Push Constants are the only reasonable way to provide the **necessary input** when using descriptor indexing.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-f2f-2023-02-16-17#push-constants-75">GPU Web F2F 2023-02-16/17</a></summary>

* D3D12: Root constants
* Metal: setVertex/Fragment/ComputeBytes
* Vulkan: push constants
* Potential proposal: passEncode.setPushConstants(offset, arraybuffer/arraybufferview)
* Very limited space: 16bytes? (root signature is constrained, push constants used by the WebGPU implementation itself).

WGSL:


```
var<push_constant> size : u32;
@fragment fn scalar() {
  _ = size;
}

struct TensorSizes {
  input1 : u32,
  input2 : u32,
}
var<push_constant> tensor_sizes : TensorSizes;
@fragment fn tensor() {
  _ = tensor_sizes;
}

```



* Jiajia: I'm making the TF.js WebGPU backend at Intel. The requirement for push constants is that they are a performance improvement and would need 64 bytes, otherwise many op can't use them.
* JQ: unary op - need at least 32 bytes. Most ops - have two inputs, one output. From TF.js POV - may need 64 bytes for push constants if you want this feature.
* MM: are push constants for performance or convenience? We can already attach data to shaders. What do these give that buffers don't?
* KG: performance.
* MM: recall I ran experiments. Performance wasn't compelling. Would need similar microbenchmark showing better performance.
* DN: Google folks have been asking for this for perf reasons, should be able to produce that.
* MM: if you show rigorous perf improvements - that's sufficient.
* CW: if we do this for perf and/or convenience reasons - there'll be discussions about when they're cleared, etc., and how much space is required in the spec. That's the scary part. Very limited space on Vulkan - competes with our impl's use of push constants. On D3D12 - the root signature max size is limited. Scary too.
* RC: and on D3D12, Dawn uses root signature to implement its own features.
* KG: somewhere it said, we could only give you 128 bytes. But if TF.js only wants 64 bytes - still valuable.
* MM: just want a compelling benchmark.
* CW: other part - what does this look like in WGSL? Bunch of restrictions. wgpu - has push constants today. Users can use in native. It looks like in the minutes (see above). Tint also has this for our internal use - not exposed to users.
* MM: don't like designing based on impls - but if design's well informed. and happens to match impls, sounds fine. Believe it can work, not a question.
* KG: no preference about syntax.
* CW: we'll wait for the data.
* KG: thanks for bringing up these topics!
* MM: excited for these additions!

</details>

**Comment by shaoboyan** [comment]
(To follow-up of this issue)
I worked out a [root constants sample](https://github.com/shaoboyan/RootConstantsPerformance) on windows (Welcome to review and comment on that sample).

The sample uses compute pipeline (flexible to adjust gpu workloads). Sample uses dispatch thread number to generate srv buffer as input and uav buffer as output. It uploads 16-512 bytes (default is 64 bytes) to uniform buffer. In shader, these uniform data is organized as `float4 scalars` , then calculate sum with data (also as `float4`) from input buffer and write the result in output buffer. There are three uniform data uploading methods in the sample: WriteBuffer, DBO and push constants. 

Sample presents average frame time and average dispatch time(*optional, needs to enable timestamp query*). Before frame started, all input buffer/staging buffer/mapping buffer has been prebaked, and the uniform data is also pre-baked in CPU array. Each frame could have several command buffer submitted and each submitted command buffer could have several dispatch commands (All configurable). Fence has been inserted between command buffer submission. Uniform data uploading happens in every dispatch.

There are two types of time: Record in CPU with [std::chrono::high_resolution_clock](https://en.cppreference.com/w/cpp/chrono/high_resolution_clock) to record average frame time and [GPU timestamp query](https://learn.microsoft.com/en-us/windows/win32/direct3d12/timing#timestamp-queries) to record per dispatch time. 
How the sample do time records :
```
// Frame started, record cpu start time
start = std::chrono::high_resolution_clock::now();

// Prepare compute command lists;
// Set Root signature, Set Descriptor Heaps
...

// Dispatch start:
// If timestamp query enabled, record start time
command_list->EndQuery(); 

// WriteBuffer: Map staging buffer --> memcpy uniform data to mapped staging buffer
// --> unmap staging buffer --> copy staging buffer to uniform buffer
// --> SetComputeRootConstantBufferView with offset calculated from dispatch count.
// Or DBO: The uniform buffer has pre-uploaded and SetComputeRootConstantBufferView
// from dispatch count
// Or RootConstants: SetComputeRoot32BitConstants with uniform data in CPU array.
...
// Set Srv and Uav
...
command_list->Dispatch();

// If timestamp query enabled, record end time
command_list->EndQuery();
// Dispatch end
...

// All dispatch finished.
command_list->ResolveQueryData();

// End command list and submit command buffer
// Insert Fence and wait for fence

end = std::chrono::high_resolution_clock::now();
// Frame end
```


Collected performance data with `[thread scale: 512 thread block * 512 thread_per_block; dispatch_num_per_frame: 512]` on Intel A770 dGPU(*To align with TFJS working model, the sample uses [__Batch mode__](https://github.com/shaoboyan/RootConstantsPerformance/tree/main#batch-mode), which mean batch all dispatch commands in the same command buffer and submit only once per frame*). Run 100 frame as warm up and then records 10 frame data.
Average Frame time, from CPU time recording:
![image](https://github.com/gpuweb/gpuweb/assets/8126135/9bfa48ad-caed-4fdf-82d3-47f36c6ff8cc)

Average Dispatch time, from GPU timestamp query:
![image](https://github.com/gpuweb/gpuweb/assets/8126135/58bb4b9e-ebda-49e3-b8b0-fe5e7adda36b)


The results shows `DBO` and `Root constants` has better performance than `WriteBuffer`. `Root constants` has nearly equal performance as (or slightly better than) `DBO`. 

But `DBO` has some limitations that developers should know all the uniform data to pre-bake it. Learned from TFJS experience, some per dispatch uniform data passes to webgpu backend before execution, and `DBO` could not help the case.

**Comment by magcius** [comment]
Going to put in a quick strawman proposal. I understand there is no practical performance benefit of push constants, but I still think the ergonomics of them are a bit nicer than UBOs. When combined with storage buffers, they provide a nice way to "bootstrap" your shader, as opposed to having to allocate and bind different UBOs or storage buffers for them.

Strawman:

You get 16 bytes, split into 4 u32's, indexed 0, 1, 2, and 3. This is enough for four indices into other buffers, which should be enough to bootstrap any proposal, while also being small enough that nobody will try to shove a matrix in there, and allowing ample room for implementations to add their own root constants.

Strawman API

`commandEncoder.setParameter(0, 0x12345678);` sets the 0'th u32 to 0x12345678.

`@parameter(0) var p0 : u32;` lets you bind the parameter in your shader.

Some optional, bikesheddable suggestions for sugar:

`@parameter(0) var p0 : vec3<u32>;` will bind parameters 0, 1, and 2 into your shader as the 0th, 1st and 2nd vector components.

`commandEncoder.setParameterFloat(0, 1.0);` will set the 0'th u32 to 0x3f800000 (the bit-encoding of the float). Use bitcast to convert to a float in the shader. This can already be done on the user side with either DataView or combining a Float32Array/Uint32Array, but this a bit more convenient.

**Comment by lcodes** [comment]
Is there any progress on this?

Even if only usable behind a feature flag, it would be useful to have.

For example, when doing hundreds to thousands of draws/dispatches when only an offset into a larger buffer changes, then it saves from having to create that many GPUBindGroup instances.

**Comment by lcodes** [comment]
In the meantime, what's the recommended approach to pass 4-bytes of data to thousands of dispatches, where each dispatch has a distinct value?

Using a storage/uniform buffer wastes 252 bytes per dispatch, due to alignment; so it's either thousands of calls to writeBuffer(), or wasting 98% of the CPU->GPU bandwidth for a single batch upload.

**Comment by shaoboyan** [comment]
Does [DBO](https://www.w3.org/TR/webgpu/#typedefdef-gpubufferdynamicoffset) or [pipelineConstantValue](https://www.w3.org/TR/webgpu/#typedefdef-gpupipelineconstantvalue) helps? 

I think this issue is still in discussion. We have benchmark (in above comments) to figure out the performance of push constants and some proposals.  For your use case, do you have any preference or thoughts on the API?

**Comment by greggman** [comment]
> In the meantime, what's the recommended approach to pass 4-bytes of data to thousands of dispatches, where each dispatch has a distinct value?

Is it possible to organize all the data in a single storage buffer and look up the per dispatch data based on the invocation id?

**Comment by magcius** [comment]
Uniform buffer offsets does not apply because of the alignment constraint. Pipeline constant values would vary per pipeline, not per dispatch. Indexing into a storage buffer would be a good idea, except there's no great way to pass the index; there's no builtin ID to identify a dispatch, and there's no multi-dispatch in WebGPU currently. This use case is a perfect example for push constants.

**Comment by devshgraphicsprogramming** [comment]
> tes per dispatch, due to alignment; so it's either thousands of calls to writeBuffer(), or wasting 98% of the CPU->GPU bandwidth for a single batch upload.

normally whatever is the equivalent of an SSBO in your API, you generally wouldn't want to rebind the buffer at different offset.

But then you need an offset for indexing :P This is why push constants are useful.

In vulkan, you can use `gl_BaseInstance` to achieve this, because then you gl_InstanceIndex gives you an already "offsetted" instanceID.

**Comment by lcodes** [comment]
@greggman The XYZ of the invocation id is already being used, but it can work as a temporary workaround with some refactoring. ie going from directly using "id.xyz" to using "id.x -> read uvec4 from storage buffer -> extract original id.xyz + offset from that uvec4" which then collapses N dispatches into a single one at the cost of additional ALU ops, memory fetches and bandwidth; but at least making savings from the original 256 bytes alignment between values.

@devshgraphicsprogramming Using `startVertex` and `startInstance` works with bitfields to pass extra data to draw calls, and its what I'm using on the rendering side, so long as the data doesn't overlap with the `vertexCount`, which means the larger the mesh or the number of instances, the less bits of data you have available to fake push constants. This would also go well with having a `drawCount` and `drawId` to `indirectDraw`. Maybe this can work with `global-invocation-id` too, in some specific cases, but not over arbitrary data. The thing with `draw` is that the count and the offset are distinct parameters for both vertex and instance, while `dispatchWorkgroups` only has the single counts.

@shaoboyan I think this is more of an API simplicity feature than a performance one, unless facing a use case where push constants make possible what otherwise requires extra memory/instructions/bandwidth.

For the API, I think making it similar to uniforms in WGSL is simple as it only introduces a new address space:

```wgsl
var<push_constants> offset : u32;
```

or

```wgsl
struct ChunkOffsets {
  offset1 : u32,
  offset2 : u32
}
var<push_constants> chunk_offsets : ChunkOffsets;
```

Then at pipeline layout creation time:

```js
device.createPipelineLayout({
  bindGroupLayouts: {...},
  pushConstants: [
    { visibility: GPUShaderStage.COMPUTE, offset: 0, size: 8 }
  ]
})
```

Finally, all that's left is a new method inside pass encoders:

```js
const pass = enc.beginComputePass();
pass.setPipeline(...);
pass.setBindGroup(0, ...);
for (let i = 0; i < numChunks; i++) {
  pass.setConstant(0, chunkOffsets, i * 2, 2); // Where chunkOffsets is a Uint32Array
  pass.dispatchWorkgroups(64, 256, 64);
}
pass.end();
```

This maps well to vulkan and others, allows using structures and typed arrays, gives control over the CPU side of how the data is broken down into parameter IDs, and works well with validation.

If performance benchmarks are important, I think they could be made against use cases like this, where the workaround using existing approaches require more work at one level or another. And in terms of hardware performance, developer productivity and application complexity. Push constants might only end up in savings for hardware performance with the largest payloads, but they're immediate savings on the other two metrics.

**Comment by shaoboyan** [comment]
I've raised a proposal for this https://github.com/gpuweb/gpuweb/pull/4612

**Comment by litherum** [comment]
> Collected performance data ... on Intel A770 dGPU

It would be useful to run this same benchmark of a variety of vendors. The investigation I [performed](https://github.com/gpuweb/gpuweb/issues/75#issuecomment-434101800
) [above](https://github.com/gpuweb/gpuweb/issues/75#issuecomment-445497161) indicates that it's a win on some devices and a loss on others.

**Comment by Kangz** [comment]
Immediates are in the spec. Closing.
